x/cron/cron.go

```diff
@@ -0,0 +1,323 @@
+specific problems. This problem is unique to this instance and this operation
+most likely succeeded on other nodes. This means that there is no way we could
+continue operating as this instance is out of sync with the rest of the
+network.
+
+%+v
+
+	`, err))
+}
+
+// tick process any number of tasks. It always returns a response and might
+// return an error. This method is similar to the Tick except it provides an
+// error. This makes it easier for the tests to check the result.
+//
+// Because this method can process multiple tasks at once, even when this
+// method returns an error instance, returned values may contain a valid
+// infromation other than the error as well. Always process returned tags and
+// validator diffs regardless the returned error value.
+func (t *Ticker) tick(ctx context.Context, db store.CacheableKVStore) ([]common.KVPair, []weave.ValidatorUpdate, error) {
+	var (
+		tags  []common.KVPair
+		vDiff []weave.ValidatorUpdate
+	)
+
+	now, err := weave.BlockTime(ctx)
+	if err != nil {
+		return tags, vDiff, errors.Wrap(err, "cannot get current time")
+	}
+	blockHeight, ok := weave.GetHeight(ctx)
+	if !ok {
+		return tags, vDiff, errors.Wrap(errors.ErrState, "cannot get current block height")
+	}
+
+	// A safety measure to not execute too many tasks in one go. Maintain
+	// an upper limit as we can process the rest of the queue in another
+	// run.
+	const maxExecuted = 50
+	for proc := 0; proc < maxExecuted; proc++ {
+		switch key, raw, err := peek(db, now); {
+		case err == nil:
+			// Each task is processed using its own cache instance
+			// to ensure changes are atomic and task processing
+			// independent.
+			cache := db.CacheWrap()
+
+			var (
+				taskTags []common.KVPair
+				taskDiff []weave.ValidatorUpdate
+			)
+			res := TaskResult{
+				Metadata:   &weave.Metadata{Schema: 1},
+				Successful: true,
+				ExecTime:   weave.AsUnixTime(now),
+				ExecHeight: blockHeight,
+			}
+
+			auth, msg, err := t.enc.UnmarshalTask(raw)
+			if err != nil {
+				res.Successful = false
+				res.Info = fmt.Sprintf("cannot unmarshal task: %+v", err)
+			} else {
+				taskCtx := withAuth(ctx, auth)
+				tx := &taskTx{msg: msg}
+				// Execute deliver using the savepoint wrapper,
+				// so that changes are applied to the cache
+				// only when the Deliver call is successful.
+				sp := utils.NewSavepoint().OnDeliver()
+				if r, err := sp.Deliver(taskCtx, cache, tx, t.hn); err != nil {
+					res.Successful = false
+					res.Info = err.Error()
+				} else {
+					taskTags = r.Tags
+					taskDiff = r.Diff
+				}
+			}
+
+			if _, err := t.results.Put(cache, key, &res); err != nil {
+				cache.Discard()
+				return tags, vDiff, errors.Wrap(err, "cannot store result")
+			}
+
+			// Remove the task from the queue as it was processed.
+			if err := cache.Delete(key); err != nil {
+				cache.Discard()
+				return tags, vDiff, errors.Wrap(err, "cannot delete task")
+			}
+			if err := cache.Write(); err != nil {
```

Comment:
Okay, here we either commit everything, or we fail.
And writing the update to task state is different than the deliver tx caching.

Well done :100:

Reply:
Team work! :man_artist: :man_astronaut: :man_scientist: